src/null-safety/migration-guide.md

```diff
@@ -1,23 +1,507 @@
-{{site.alert.end}}
+This page describes how and when to migrate your code to [null safety][].
+Here are the basic steps for migrating each package that you own:
 
-For more information, see the [null safety homepage][].
+1. [**Wait**](#step1-wait) for the packages
+   that you depend on to migrate.
+2. [**Migrate**](#step2-migrate) your package's code,
+   preferably using the interactive migration tool.
+3. [**Test**](#step3-test) to make sure your changes work.
+4. If the package is already on pub.dev,
+   [**publish**](#step4-publish) the null-safe version
+   as a **prerelease** version.
 
-[null safety homepage]: /null-safety
-[Dart announcements]: {{site.group}}/d/forum/announce
-[Dart blog]: https://medium.com/dartlang
+{{ site.alert.info }}
+  **Migrating an app is the same as migrating a package.**
+  Before fully migrating an app,
+  consider waiting until null safety is in a stable release.
+  In the meantime, you can learn about null safety,
+  perhaps adding migration hints to your app
+  or refactoring your app code.
+{{ site.alert.end }}
 
+[null safety]: /null-safety
+
+![Screenshot of migration tool](/null-safety/migration-tool.png)
+_The interactive migration tool can simplify migration to null safety._
+
+
+## 1. Wait to migrate {#step1-wait}
+
+We strongly recommend migrating code in order, 
+with the leaves of the dependency graph being migrated first.
+For example, if package C depends on package B, which depends on package A,
+then A should be migrated to null safety first, then B, then C.
+
+![Illustration of C/B/A sentence](/null-safety/null-safety-migration-order.png){:width="454px"}<br>
+
+Although you _can_ migrate before your dependencies support null safety,
+you might have to change your code when your dependencies migrate.
+For example, if you predict that a function will take a nullable parameter but
+the package migrates it to be non-nullable,
+then passing a nullable argument becomes a compile error.
+
+This section tells you how to
+check your package's dependencies
+by temporarily opting into null safety
+and using the `dart pub outdated` command in null-safety mode.
+
+{{ site.alert.warn }}
+  These instructions assume your code is under **source control**,
+  so that you can easily undo any changes.
+{{ site.alert.end }}
+
+### Switch to a 2.12 beta release
+
+**Switch to the latest beta release**
+of either the Dart SDK or the Flutter SDK.
+How you get the latest beta release depends on whether
+you use the Flutter SDK:
+
+* If you use the Flutter SDK,
+  switch to the beta channel:
+
+  ```terminal
+$ flutter channel beta
+$ flutter upgrade
+```
+* Otherwise, download a beta release
+  from the [Dart SDK archive][].
+
+[Dart SDK archive]: /tools/sdk/archive#beta-channel
+
+{{ site.alert.warn }}
+  If you return to working on production code, remember to
+  **switch back to a stable release**
+  (for example, by running `flutter channel stable`).
+{{ site.alert.end }}
+
+
+### Opt into null safety
```

Comment:
@jonasfj @munificent is this needed here while checking if dependencies are ready, or could that be done with a 2.10 SDK? I guess it might not be able to resolve deps with bad SDK constraints, but generally it should not be needed, right?

Here the section is in the staged content:
https://kw-www-dartlang-1.firebaseapp.com/null-safety/migration-guide#opt-into-null-safety